native/membrane_videocompositor_opengl_rust/src/lib.rs

```diff
@@ -0,0 +1,246 @@
+    egl::GREEN_SIZE,    8,
+    egl::BLUE_SIZE,     8,
+    egl::CONFORMANT,    egl::OPENGL_ES3_BIT,
+    egl::NONE
+  ];
+
+  let config = egl
+    .choose_first_config(display, &attributes)
+    .nif_err()?
+    .expect("Got no compatible config");
+
+  #[rustfmt::skip]
+  let attributes = [
+    egl::CONTEXT_MAJOR_VERSION, 3,
+    egl::CONTEXT_MINOR_VERSION, 0,
+    egl::NONE,
+  ];
+
+  let context = egl
+    .create_context(display, config, None, &attributes)
+    .nif_err()?;
+
+  egl.make_current(display, None, None, Some(context)).nif_err()?;
+
+  gl::load(|name| {
+    egl
+      .get_proc_address(name)
+      .expect("Can't find a GLES procedure") as *const std::ffi::c_void
+  });
+
+  unsafe { gl::ClearColor(0.0, 0.0, 0.0, 1.0) }
+
+  let vertex_shader_code = include_str!("shaders/vertex.glsl");
+  let fragment_shader_code = include_str!("shaders/fragment.glsl");
+
+  let shader_program = ShaderProgram::new(vertex_shader_code, fragment_shader_code)?;
+  let mut scene = Scene::new(out_video.width, out_video.height, shader_program);
+  scene.add_video(
+    scene::VideoPlacementTemplate {
+      top_right: (1.0, 1.0),
+      top_left: (-1.0, 1.0),
+      bot_left: (-1.0, 0.0),
+      bot_right: (1.0, 0.0),
+      z_value: 0.0
+    },
+    first_video.width, second_video.height
+  );
+
+  scene.add_video(
+    scene::VideoPlacementTemplate {
+      top_right: (1.0, 0.0),
+      top_left: (-1.0, 0.0),
+      bot_left: (-1.0, -1.0),
+      bot_right: (1.0, -1.0),
+      z_value: 0.0
+    },
+    first_video.width, second_video.height
+  );
+
+
+
+  egl.make_current(display, None, None, None).nif_err()?;
+  Ok((atoms::ok(), ResourceArc::new(State::new(display, context, scene))))
+}
+
+#[rustler::nif]
+/// Join two frames passed as [binaries](rustler::Binary). This function is intended to only be called by Elixir.
+fn join_frames<'a>(env: rustler::Env<'a>, state: rustler::ResourceArc<State>, upper: rustler::Binary, lower: rustler::Binary) -> Result<(rustler::Atom, rustler::Term<'a>), rustler::Error> {
+  let ctx: BoundContext = state.bind_context().nif_err()?;
+  // for some reason VS Code can't suggest stuff correctly until 
+  // I forward all of this into a different function. It's inlined and thusly free
+  join_frames_fwd(env, &ctx, upper, lower)
+}
+
+#[inline(always)]
+#[doc(hidden)]
+fn join_frames_fwd<'a>(env: rustler::Env<'a>, ctx: &BoundContext, upper: rustler::Binary, lower: rustler::Binary)  -> Result<(rustler::Atom, rustler::Term<'a>), rustler::Error> {
+  ctx.scene.upload_texture(0, upper.as_slice());
+  ctx.scene.upload_texture(1, lower.as_slice());
+
+  let mut binary = rustler::OwnedBinary::new(ctx.scene.out_width() * ctx.scene.out_height() * 3 / 2).unwrap();
+  ctx.scene.draw_into(binary.as_mut_slice());
+  Ok((atoms::ok(), binary.release(env).to_term(env)))
+}
+
+
+trait ResultExt<T> {
+  /// Convert `T` into a [Result] with a [rustler::Error].
+  fn nif_err(self) -> Result<T, rustler::Error>;
+}
```

Comment:
That's smart, but actually I think I would generalise this over any Error-like types, instead of being focused just on `Result<T, E>` wrapper.

If we hadn't have orphan rule in Rust, you could implement `impl From<egl::Error> for rustler::Error`. For now, you can define a `From`-like trait which is `rustler::Error` specific and put impls for all interesting structs.

Reply:
You mean I should do something like this:

```rs
trait ToRustlerError {
    fn convert(self) -> rustler::Error;
}

impl ToRustlerError for egl::Error {
    ...
}
```
and then `.map_err(|e| e.convert())` everywhere?

Reply:
Actually, I think I came up with a potentially better solution while working on error handling. I can have my own `CompositorError` type, which implements `From<egl::Error>` (along with a couple more `From<T>`s for all kinds of errors that can arise in the project) and implement `From<CompositorError>` for `rustler::Error`. Then error handling becomes very nice, since we can just use the `?` operator virtually everywhere we want. What do you think of that?

Reply:
Yup, that sounds good!

Reply:
I'll resolve this conversation then, since I already have a PR implementing this idea.